pkg/strategy/factorzoo/factors/amplitude_volume_divergence.go

```diff
@@ -7,29 +7,56 @@ import (
 
 	"github.com/c9s/bbgo/pkg/indicator"
 	"github.com/c9s/bbgo/pkg/types"
+
+	log "github.com/sirupsen/logrus"
 )
 
 var zeroTime time.Time
 
 type KLineValueMapper func(k types.KLine) float64
 
-//go:generate callbackgen -type Correlation
-type Correlation struct {
+//go:generate callbackgen -type AVD
+type AVD struct {
+	types.SeriesBase
 	types.IntervalWindow
-	Values  types.Float64Slice
+
+	// Values
+	Values    types.Float64Slice
+	LastValue float64
+
 	EndTime time.Time
 
-	UpdateCallbacks []func(value float64)
+	UpdateCallbacks []func(val float64)
 }
 
-func (inc *Correlation) Last() float64 {
-	if len(inc.Values) == 0 {
-		return 0.0
+func (inc *AVD) Index(i int) float64 {
+	if inc.Values == nil {
+		return 0
 	}
-	return inc.Values[len(inc.Values)-1]
+	return inc.Values.Index(i)
 }
 
-func (inc *Correlation) CalculateAndUpdate(klines []types.KLine) {
+func (inc *AVD) Last() float64 {
+	if inc.Values.Length() == 0 {
+		return 0
+	}
+	return inc.Values.Last()
+}
+
+func (inc *AVD) Length() int {
+	if inc.Values == nil {
+		return 0
+	}
+	return inc.Values.Length()
+}
+
+//var _ types.SeriesExtend = &AVD{}
+
+func (inc *AVD) Update(klines []types.KLine) {
```

Comment:
this function should be called `CalculateAndUpdate`, check the latest EWMA indicator's interface:

BindK()
PushK()
LoadK()